test/src/loaders/workspace_uploader/workspace_uploader_test.py

```diff
@@ -1,5 +1,302 @@
+        target_file_path = target_dir_path.joinpath(assembly_name)
+        target_file_path.touch()
+        new_dir_path = collection_source_dir.joinpath(assembly_dir_name)
+        os.symlink(
+            target_dir_path.resolve(), new_dir_path.resolve(), target_is_directory=True
+        )
+        assembly_dirs.append(str(new_dir_path))
+        target_files.append(str(target_file_path))
+
+    ParamsTuple = namedtuple(
+        "ParamsTuple",
+        [
+            "tmp_dir",
+            "sourcedata_dir",
+            "collection_source_dir",
+            "assembly_dirs",
+            "target_files",
+        ],
+    )
+
+    yield ParamsTuple(
+        tmp_dir, sourcedata_dir, collection_source_dir, assembly_dirs, target_files
+    )
+
+    shutil.rmtree(tmp_dir, ignore_errors=True)
+
+
+def test_get_yaml_file_path(setup_and_teardown):
+    params = setup_and_teardown
+    assembly_dir = params.assembly_dirs[0]
+    yaml_path = workspace_uploader._get_yaml_file_path(assembly_dir)
+
+    expected_yaml_path = os.path.join(assembly_dir, workspace_uploader.UPLOADED_YAML)
+    assert expected_yaml_path == yaml_path
+
+
+def test_get_source_file(setup_and_teardown):
+    params = setup_and_teardown
+    assembly_dir, assembly_name = params.assembly_dirs[0], ASSEMBLY_NAMES[0]
+    file_path = workspace_uploader._get_source_file(assembly_dir, assembly_name)
+
+    expected_target_file_path = os.path.abspath(params.target_files[0])
+    assert expected_target_file_path == file_path
+
+
+def test_read_upload_status_yaml_file(setup_and_teardown):
+    params = setup_and_teardown
+    assembly_dir = params.assembly_dirs[0]
+    assembly_name = ASSEMBLY_NAMES[0]
+
+    upload_env_key = "CI"
+    workspace_id = 12345
+
+    # test empty yaml file in assembly_dir
+    data, uploaded = workspace_uploader._read_upload_status_yaml_file(
+        upload_env_key, workspace_id, assembly_dir, assembly_name
+    )
+
+    expected_data = {upload_env_key: {workspace_id: dict()}}
+
+    assert not uploaded
+    assert expected_data == data
+
+
+def test_update_upload_status_yaml_file(setup_and_teardown):
+    params = setup_and_teardown
+    upload_env_key = "CI"
+    upa = "12345_58_1"
+    workspace_id = 12345
+    assembly_dir = params.assembly_dirs[0]
+    assembly_name = ASSEMBLY_NAMES[0]
+
+    workspace_uploader._update_upload_status_yaml_file(
+        upload_env_key, workspace_id, upa, assembly_dir, assembly_name
+    )
+    data, uploaded = workspace_uploader._read_upload_status_yaml_file(
+        upload_env_key, workspace_id, assembly_dir, assembly_name
+    )
+
+    expected_data = {
+        upload_env_key: {workspace_id: {"file_name": assembly_name, "upa": upa}}
+    }
+
+    assert uploaded
+    assert expected_data == data
+
+    with pytest.raises(ValueError, match=f"already exists in workspace"):
+        workspace_uploader._update_upload_status_yaml_file(
+            upload_env_key, workspace_id, upa, assembly_dir, assembly_name
+        )
```

Comment:
You can inline a lot of this stuff - this is just an example

```suggestion
    assembly_dir = params.assembly_dirs[0]
    assembly_name = ASSEMBLY_NAMES[0]

    workspace_uploader._update_upload_status_yaml_file(
        "PROD", 42, "34_2_7", assembly_dir, assembly_name
    )
    data, uploaded = workspace_uploader._read_upload_status_yaml_file(
        "PROD", 42, assembly_dir, assembly_name
    )

    expected_data = {
        "PROD": {42: {"file_name": assembly_name, "upa": "34_2_7"}}
    }

    assert uploaded
    assert expected_data == data

    with pytest.raises(ValueError, match=f"already exists in workspace"):
        workspace_uploader._update_upload_status_yaml_file(
            "PROD", 42, "34_2_7", assembly_dir, assembly_name
        )
```

Reply:
What are the benefits doing this? I kinda feel inline is more error-prone.

Reply:
* The big on is that you're now testing based on equality rather than identity. This is more important for classes but it's a good habit to get into.
* fewer test lines
* easier to read the tests naturally
* easier to check that your tests work by forcing them to fail

Reply:
Updated!

Reply:
:+1: